File: src/oca/java/test/TemplateTest.java
/*******************************************************************************
 * Copyright 2002-2012, OpenNebula Project Leads (OpenNebula.org)
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 ******************************************************************************/
import static org.junit.Assert.assertTrue;

import org.junit.After;
import org.junit.AfterClass;
import org.junit.Before;
import org.junit.BeforeClass;
import org.junit.Test;
import org.opennebula.client.Client;
import org.opennebula.client.OneResponse;
import org.opennebula.client.group.Group;
import org.opennebula.client.template.*;
import org.opennebula.client.user.User;
import org.opennebula.client.vm.VirtualMachine;



public class TemplateTest
{

    private static Template template;
    private static TemplatePool templatePool;

    private static Client client;

    private static OneResponse res;
    private static String name = "new_test_template";


    private static String template_str =
        "NAME = \"" + name + "\"\n" +
        "ATT1 = \"VAL1\"\n" +
        "ATT2 = \"VAL2\"";

    /**
     * @throws java.lang.Exception
     */
    @BeforeClass
    public static void setUpBeforeClass() throws Exception
    {
        client          = new Client();
        templatePool    = new TemplatePool(client);
    }

    /**
     * @throws java.lang.Exception
     */
    @AfterClass
    public static void tearDownAfterClass() throws Exception
    {
    }

    /**
     * @throws java.lang.Exception
     */
    @Before
    public void setUp() throws Exception
    {
        res = Template.allocate(client, template_str);

        int oid = res.isError() ? -1 : Integer.parseInt(res.getMessage());
        template = new Template(oid, client);
    }

    /**
     * @throws java.lang.Exception
     */
    @After
    public void tearDown() throws Exception
    {
        template.delete();
    }

    @Test
    public void allocate()
    {
        template.delete();

        res = Template.allocate(client, template_str);
        assertTrue( !res.isError() );

        int oid = res.isError() ? -1 : Integer.parseInt(res.getMessage());
        template = new Template(oid, client);


        templatePool.info();

        boolean found = false;
        for(Template temp : templatePool)
        {
            found = found || temp.getName().equals(name);
        }

        assertTrue( found );
    }

    @Test
    public void info()
    {
        res = template.info();
        assertTrue( !res.isError() );

//        assertTrue( template.getId().equals("0") );
//        assertTrue( template.id() == 0 );
        assertTrue( template.getName().equals(name) );
    }

    @Test
    public void update()
    {
        res = template.info();
        assertTrue( !res.isError() );

        assertTrue( template.xpath("TEMPLATE/ATT1").equals( "VAL1" ) );
        assertTrue( template.xpath("TEMPLATE/ATT2").equals( "VAL2" ) );

        String new_template =   "ATT2 = NEW_VAL\n" +
                                "ATT3 = VAL3";

        res = template.update(new_template);
        assertTrue( !res.isError() );


        res = template.info();
        assertTrue( !res.isError() );
        assertTrue( template.xpath("TEMPLATE/ATT1").equals( "" ) );
        assertTrue( template.xpath("TEMPLATE/ATT2").equals( "NEW_VAL" ) );
        assertTrue( template.xpath("TEMPLATE/ATT3").equals( "VAL3" ) );
    }

    @Test
    public void publish()
    {
        res = template.publish();
        assertTrue( !res.isError() );

        template.info();
        assertTrue( template.xpath("PERMISSIONS/GROUP_U").equals("1") );
    }

    @Test
    public void unpublish()
    {
        res = template.unpublish();
        assertTrue( !res.isError() );

        template.info();
        assertTrue( template.xpath("PERMISSIONS/GROUP_U").equals("0") );
    }

    @Test
    public void chmod()
    {
        res = template.info();
        assertTrue( !res.isError() );

        String owner_a = template.xpath("PERMISSIONS/OWNER_A");
        String group_a = template.xpath("PERMISSIONS/GROUP_A");

        res = template.chmod(0, 1, -1, 1, 0, -1, 1, 1, 0);
        assertTrue( !res.isError() );

        res = template.info();
        assertTrue( !res.isError() );

        assertTrue( template.xpath("PERMISSIONS/OWNER_U").equals("0") );
        assertTrue( template.xpath("PERMISSIONS/OWNER_M").equals("1") );
        assertTrue( template.xpath("PERMISSIONS/OWNER_A").equals(owner_a) );
        assertTrue( template.xpath("PERMISSIONS/GROUP_U").equals("1") );
        assertTrue( template.xpath("PERMISSIONS/GROUP_M").equals("0") );
        assertTrue( template.xpath("PERMISSIONS/GROUP_A").equals(group_a) );
        assertTrue( template.xpath("PERMISSIONS/OTHER_U").equals("1") );
        assertTrue( template.xpath("PERMISSIONS/OTHER_M").equals("1") );
        assertTrue( template.xpath("PERMISSIONS/OTHER_A").equals("0") );
    }

    @Test
    public void chmod_octet()
    {
        res = template.info();
        assertTrue( !res.isError() );

        res = template.chmod(640);
        assertTrue( !res.isError() );

        res = template.info();
        assertTrue( !res.isError() );

        assertTrue( template.xpath("PERMISSIONS/OWNER_U").equals("1") );
        assertTrue( template.xpath("PERMISSIONS/OWNER_M").equals("1") );
        assertTrue( template.xpath("PERMISSIONS/OWNER_A").equals("0") );
        assertTrue( template.xpath("PERMISSIONS/GROUP_U").equals("1") );
        assertTrue( template.xpath("PERMISSIONS/GROUP_M").equals("0") );
        assertTrue( template.xpath("PERMISSIONS/GROUP_A").equals("0") );
        assertTrue( template.xpath("PERMISSIONS/OTHER_U").equals("0") );
        assertTrue( template.xpath("PERMISSIONS/OTHER_M").equals("0") );
        assertTrue( template.xpath("PERMISSIONS/OTHER_A").equals("0") );

        res = template.chmod("147");
        assertTrue( !res.isError() );

        res = template.info();
        assertTrue( !res.isError() );

        assertTrue( template.xpath("PERMISSIONS/OWNER_U").equals("0") );
        assertTrue( template.xpath("PERMISSIONS/OWNER_M").equals("0") );
        assertTrue( template.xpath("PERMISSIONS/OWNER_A").equals("1") );
        assertTrue( template.xpath("PERMISSIONS/GROUP_U").equals("1") );
        assertTrue( template.xpath("PERMISSIONS/GROUP_M").equals("0") );
        assertTrue( template.xpath("PERMISSIONS/GROUP_A").equals("0") );
        assertTrue( template.xpath("PERMISSIONS/OTHER_U").equals("1") );
        assertTrue( template.xpath("PERMISSIONS/OTHER_M").equals("1") );
        assertTrue( template.xpath("PERMISSIONS/OTHER_A").equals("1") );
    }
    
    @Test
    public void attributes()
    {
        res = template.info();
        assertTrue( !res.isError() );

//        assertTrue( template.xpath("ID").equals("0") );
        assertTrue( template.xpath("NAME").equals(name) );
    }

    @Test
    public void delete()
    {
        res = template.delete();
        assertTrue( !res.isError() );

        res = template.info();
        assertTrue( res.isError() );
    }

    @Test
    public void chown()
    {
        // Create a new User and Group
        res = User.allocate(client, "template_test_user", "password");
        assertTrue( !res.isError() );

        int uid = Integer.parseInt(res.getMessage());

        res = Group.allocate(client, "template_test_group");
        assertTrue( !res.isError() );

        int gid = Integer.parseInt(res.getMessage());

        res = template.info();
        assertTrue( !res.isError() );

        assertTrue( template.uid() == 0 );
        assertTrue( template.gid() == 0 );

        res = template.chown(uid, gid);
        assertTrue( !res.isError() );

        res = template.info();
        assertTrue( !res.isError() );

        assertTrue( template.uid() == uid );
        assertTrue( template.gid() == gid );

        res = template.chgrp(0);

        res = template.info();
        assertTrue( !res.isError() );

        assertTrue( template.uid() == uid );
        assertTrue( template.gid() == 0 );
    }

    @Test
    public void instantiate()
    {
        res = template.instantiate("new_vm_name");
        assertTrue( !res.isError() );

        int vm_id = Integer.parseInt(res.getMessage());
        VirtualMachine vm = new VirtualMachine(vm_id, client);

        res = vm.info();
        assertTrue( !res.isError() );

        assertTrue( vm.getName().equals( "new_vm_name" ) );
    }
}

File: src/oca/go/src/goca/template_test.go
/* -------------------------------------------------------------------------- */
/* Copyright 2002-2025, OpenNebula Project, OpenNebula Systems                */
/*                                                                            */
/* Licensed under the Apache License, Version 2.0 (the "License"); you may    */
/* not use this file except in compliance with the License. You may obtain    */
/* a copy of the License at                                                   */
/*                                                                            */
/* http://www.apache.org/licenses/LICENSE-2.0                                 */
/*                                                                            */
/* Unless required by applicable law or agreed to in writing, software        */
/* distributed under the License is distributed on an "AS IS" BASIS,          */
/* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.   */
/* See the License for the specific language governing permissions and        */
/* limitations under the License.                                             */
/*--------------------------------------------------------------------------- */

package goca

import (
	. "gopkg.in/check.v1"

	"github.com/OpenNebula/one/src/oca/go/src/goca/schemas/vm"
	"github.com/OpenNebula/one/src/oca/go/src/goca/schemas/vm/keys"
	"github.com/OpenNebula/one/src/oca/go/src/goca/schemas/shared"
)

type TemplateSuite struct {
	ID       int
}

var _ = Suite(&TemplateSuite{})

func (s *TemplateSuite) SetUpTest(c *C) {
	templateName := GenName("template")

	// Create template
	tpl := vm.NewTemplate()
	tpl.Add(keys.Name, templateName)
	tpl.CPU(1).Memory(64)

	id, err := testCtrl.Templates().Create(tpl.String())
	c.Assert(err, IsNil)

	s.ID = id
}

func (s *TemplateSuite) TearDownTest(c *C) {
	// Delete tmpl
	testCtrl.Template(s.ID).Delete()
}

func (s *TemplateSuite) TestGetByNameAndID(c *C) {
	// Get tmpl by ID
	tmplC := testCtrl.Template(s.ID)
	tmpl, err := tmplC.Info(false, false)
	c.Assert(err, IsNil)
	c.Assert(tmpl.ID, Equals, s.ID)

	// Get tmpl by Name
	id, err := testCtrl.Templates().ByName(tmpl.Name)
	c.Assert(err, IsNil)
	c.Assert(tmpl.ID, Equals, id)
}

func (s *TemplateSuite) TestUpdate(c *C) {
	// Update
	tmplC := testCtrl.Template(s.ID)
	err := tmplC.Update("A=B", 1)
	c.Assert(err, IsNil)

	tmpl, err := tmplC.Info(false, false)
	c.Assert(err, IsNil)

	val, err := tmpl.Template.GetStr("A")
	c.Assert(err, IsNil)
	c.Assert(val, Equals, "B")
}

func (s *TemplateSuite) TestRename(c *C) {
	tmplC := testCtrl.Template(s.ID)
	tmplC.Rename("new_name")

	tmpl, err := tmplC.Info(false, false)
	c.Assert(err, IsNil)
	c.Assert(tmpl.Name, Equals, "new_name");
}

func (s *TemplateSuite) TestChmod(c *C) {
	new_permissions := shared.Permissions{1, 1, 1, 1, 1, 1, 1, 1, 1}

	tmplC := testCtrl.Template(s.ID)

	err := tmplC.Chmod(new_permissions)

	c.Assert(err, IsNil)

	tmpl, err := tmplC.Info(false, false)

	c.Assert(err, IsNil)
	c.Assert(*tmpl.Permissions, Equals, new_permissions);
}

func (s *TemplateSuite) TestChown(c *C) {
	// Test only if the call exists, no real change
	tmplC := testCtrl.Template(s.ID)
	err := tmplC.Chown(1, 1)

	c.Assert(err, IsNil)

	tmpl, err := tmplC.Info(false, false)
	c.Assert(err, IsNil)
	c.Assert(tmpl.UID, Equals, 1);
	c.Assert(tmpl.GID, Equals, 1);
}

func (s *TemplateSuite) TestLock(c *C) {
	// Lock
	tmplC := testCtrl.Template(s.ID)
	err := tmplC.Lock(shared.LockUse)
	c.Assert(err, IsNil)

	tmpl, err := tmplC.Info(false, false)
	c.Assert(err, IsNil)
	c.Assert(tmpl.LockInfos.Locked, Equals, 1);

	// Unlock
	err = tmplC.Unlock()
	c.Assert(err, IsNil)

	tmpl, err = tmplC.Info(false, false)
	c.Assert(err, IsNil)
	c.Assert(tmpl.LockInfos, IsNil)
}

func (s *TemplateSuite) TestClone(c *C) {
	name := GenName("cloned_tmpl")
	tmplC := testCtrl.Template(s.ID)
	err := tmplC.Clone(name, false)
	c.Assert(err, IsNil)

	// Delete cloned tmpl
	id, err := testCtrl.Templates().ByName(name)
	c.Assert(err, IsNil)

	testCtrl.Template(id).Delete()
}

func (s *TemplateSuite) TestInstantiate(c *C) {
	// Get template by ID
	tmplC := testCtrl.Template(s.ID)

	// Instantiate(name string, pending bool, extra string) (uint, error)
	vmid, err := tmplC.Instantiate("", false, "", false)
	c.Assert(err, IsNil)

	err = testCtrl.VM(vmid).Terminate()
	c.Assert(err, IsNil)
}
